Add lis2du12 accelerometer driver

avisconti opened the pull request:
The LIS2DU12 is a linear ultralow power 3-axis accelerometer sensor.
This driver is based on stmemsc HAL i/f v2.3

avisconti committed:
drivers/sensor: add support to LIS2DU12 accelerometer

The LIS2DU12 is a linear 3-axis accelerometer with advanced digital
functions whose MEMS and ASIC have been expressly designed to build
an outstanding ultralow-power architecture in which the anti-aliasing
filter operates with a current consumption among the lowest in the
market.
This driver is based on stmemsc HAL i/f v2.3

https://www.st.com/en/datasheet/lis2du12.pdf

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
boards: arm: sensortile_box_pro: extend with lis2du12

Extend sensortile_box_pro with lis2du12 accelerometer.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
sample: board: sensortile_box_pro: extend with lis2du12 accelerometer

Extend sensor sample reading also lis2du12 accelerometer data.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

kartben commented on dts/bindings/sensor/st,lis2du12-common.yaml:
```diff
@@ -0,0 +1,97 @@
+    type: int
+    default: 0
+    description: |
+      Range in g. Default is power-up configuration.
```
nit: it would be great to consider doing a mass replace of s/g/G/ as there are plenty of places where the unit is mispelled (applies to "mg" just below too). It concerns mostly the ST accelerometers but a few others as well.

avisconti replied:
> nit: it would be great to consider doing a mass replace of s/g/G/ as there are plenty of places where the unit is mispelled (applies to "mg" just below too). It concerns mostly the ST accelerometers but a few others as well.

@kartben in all out datasheet I read g/mg instead of G/mG. I'm failing to find with google a clear advise on which one to use...

kartben replied:
Mmmm yeah my bad. It would seem that "g" is indeed the way to go. I guess people preferred risking a confusion with "grams" than a confusion with G the gravitational constant. It hurts my brain, but definitely not something that needs fixing in the end, sorry for the noise!

avisconti commented:
Any feedback here?

fabiobaltieri commented on drivers/sensor/lis2du12/lis2du12.h:
```diff
@@ -0,0 +1,102 @@
+#include <zephyr/drivers/i2c.h>
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) */
+
+#define LIS2DU12_EN_BIT					0x01
+#define LIS2DU12_DIS_BIT					0x00
```
the indentation looks funny, I'd sugget to just use a space

```c
#define LIS2DU12_EN_BIT	 0x01
#define LIS2DU12_DIS_BIT 0x00
```

no need to make it fancy, same below

avisconti replied:
mmmh, I think that maybe tabs are better. But I'd agree to shorten it a bit and align with below

fabiobaltieri replied:
Or just a space... it's just to, I wouldn't bother.

fabiobaltieri commented on boards/arm/sensortile_box_pro/sensortile_box_pro.dts:
```diff
@@ -180,6 +180,16 @@
+
+	lis2du12: lis2du12@1 {
+		compatible = "st,lis2du12";
+		spi-max-frequency = <DT_FREQ_M(10)>; /* 10 MHz */
```
drop the comment, the code is meaningful enough without it, it's just going to be copy pasted around and go out of date

avisconti replied:
ok, makes sense

avisconti commented on drivers/sensor/lis2du12/lis2du12_trigger.c:
```diff
@@ -0,0 +1,258 @@
+			return lis2du12_enable_xl_int(dev, LIS2DU12_DIS_BIT);
+		}
+	}
+#if defined(CONFIG_LIS2DU12_ENABLE_TEMP)
+	else if (trig->chan == SENSOR_CHAN_DIE_TEMP) {
```
OK, converted to a switch. Anyway, I currently removed the die temperature sensor part. It is not yet supported in the stmemsc driver. I'll take care of it and in the next stmemsc module version I will re-add it.

avisconti commented on drivers/sensor/lis2du12/lis2du12_trigger.c:
```diff
@@ -0,0 +1,258 @@
+
+	ret = gpio_pin_configure_dt(lis2du12->drdy_gpio, GPIO_INPUT);
+	if (ret < 0) {
+		LOG_DBG("Could not configure gpio");
```
just did it for all the logs now

avisconti commented on drivers/sensor/lis2du12/lis2du12.c:
```diff
@@ -0,0 +1,516 @@
+
+LOG_MODULE_REGISTER(LIS2DU12, CONFIG_SENSOR_LOG_LEVEL);
+
+static const float lis2du12_odr_map[14] = {
```
Currently no.
I might add LIS2DU12_ODR_NUM in following stmemsc struct, even if I don't like it. But anyway in the future, because I need to change/integrate the stmemsc module.

```
typedef struct {
  enum {
    LIS2DU12_OFF               = 0x00, /* in power down */
    LIS2DU12_1Hz6_ULP          = 0x01, /* @1Hz6(ultralow power) */
    LIS2DU12_3Hz_ULP           = 0x02, /* @3Hz (ultralow power) */
    LIS2DU12_6Hz_ULP           = 0x03, /* @6Hz (ultralow power) */
    LIS2DU12_6Hz               = 0x04, /* @6Hz (normal) */
    LIS2DU12_12Hz5             = 0x05, /* @12Hz5 (normal) */
    LIS2DU12_25Hz              = 0x06, /* @25Hz  (normal) */
    LIS2DU12_50Hz              = 0x07, /* @50Hz  (normal) */
    LIS2DU12_100Hz             = 0x08, /* @100Hz (normal) */
    LIS2DU12_200Hz             = 0x09, /* @200Hz (normal) */
    LIS2DU12_400Hz             = 0x0A, /* @400Hz (normal) */
    LIS2DU12_800Hz             = 0x0B, /* @800Hz (normal) */
    LIS2DU12_TRIG_PIN          = 0x0E, /* Single-shot high latency by INT2 */
    LIS2DU12_TRIG_SW           = 0x0F, /* Single-shot high latency by IF */
    LIS2DU12_ODR_NUM    = 14         /* number of sensor rate frequencies */
  } odr;                                                                                                                                                                                       
  enum {
    LIS2DU12_2g   = 0, 
    LIS2DU12_4g   = 1, 
    LIS2DU12_8g   = 2, 
    LIS2DU12_16g  = 3, 
  } fs;
  enum {
    LIS2DU12_ODR_div_2   = 0, 
    LIS2DU12_ODR_div_4   = 1, 
    LIS2DU12_ODR_div_8   = 2, 
    LIS2DU12_ODR_div_16  = 3, 
  } bw;
} lis2du12_md_t;
```

fabiobaltieri commented on drivers/sensor/lis2du12/Kconfig:
```diff
@@ -0,0 +1,61 @@
+	select HAS_STMEMSC
+	select USE_STDC_LIS2DU12
+	help
+	  Enable driver for LIS2DU12 smartxl
+	  sensor.
```
Nit: this fits on a single
line.

fabiobaltieri commented on drivers/sensor/lis2du12/Kconfig:
```diff
@@ -0,0 +1,61 @@
+	  Enable driver for LIS2DU12 smartxl
+	  sensor.
+
+if LIS2DU12
+
+choice LIS2DU12_TRIGGER_MODE
+	prompt "Trigger mode"
+	help
+	  Specify the type of triggering to be used by the driver.
+
+config LIS2DU12_TRIGGER_NONE
+	bool "No trigger"
+
+config LIS2DU12_TRIGGER_GLOBAL_THREAD
+	bool "Use global thread"
+	depends on GPIO
+	select LIS2DU12_TRIGGER
+
+config LIS2DU12_TRIGGER_OWN_THREAD
+	bool "Use own thread"
+	depends on GPIO
+	select LIS2DU12_TRIGGER
+
+endchoice
```
unrelated to this PR but let me pitch the idea to you: this pattern of having an option for how to handle asyncronous events (thread, workqueue) is copypasted everywhere. How would you feel about having a Kconfig template to define these options (like the log one, the one you use with `source "subsys/logging/Kconfig.template.log_config"`) and then common set of of macro to prepare and use the support data for the selected operating mode? (thread, workqueue, syncronous or none)

Haven't quite tried to do it but I think it could remove lot of duplicated code.

avisconti replied:
Yes, looks a great idea. Let me explore more on that.

fabiobaltieri replied:
Thanks!

avisconti replied:
ok, take a look just to see I did what you wanted to (not 100% sure)

fabiobaltieri replied:
brilliant, yeah I was actually thinking about also having a set of macros to dedup the code as well but this is a brilliant start and a step towards it, thanks for doing it, let's think about the code in followups

@MaureenHelm @teburd what do you think?

teburd replied:
It would be great, thought to be frank I'd rather see us move in the newer sensor read API approach which avoids the need for work queues/threads at all

fabiobaltieri commented on drivers/sensor/lis2du12/lis2du12.c:
```diff
@@ -0,0 +1,516 @@
+}
+
+static int lis2du12_attr_set(const struct device *dev,
+			    enum sensor_channel chan,
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
```
still applicable

avisconti replied:
sorry, missed it...

fabiobaltieri commented on drivers/sensor/lis2du12/lis2du12.c:
```diff
@@ -0,0 +1,516 @@
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return lis2du12_accel_odr_set(dev, val->val1);
+	default:
+		LOG_DBG("Accel attribute not supported.");
```
I think I would do warning here and keep errors for stuff that breaks the driver permanently, like initialization errors.

avisconti committed:
drivers/sensor: add Kconfig.trigger_template

Add Kconfig.trigger_template to allow an extensive re-use of
trigger configuration inside all sensor drivers.

This template must be included as in the following example:

    module = LSM6DSO
    thread_priority = 10
    thread_stack_size = 1024
    source "drivers/sensor/Kconfig.trigger_template"

Signed-off-by: Armando Visconti <armando.visconti@st.com>
